### PR TITLE
native-posix: Add default CONFIG_CPU_HAS_FPU

### PR DESCRIPTION
**Pull request by yperess**
Add the ability to run native-posix tests using the FPU.

Signed-off-by: Yuval Peress <peress@google.com>

### CODE REVIEW

File: boards/posix/native_posix/Kconfig.defconfig [hunk 1]
```diff
@@ -15,6 +15,9 @@ config BOARD
 	default "native_posix_64" if BOARD_NATIVE_POSIX_64BIT
 	default "native_posix"
 
+config CPU_HAS_FPU
+        default y
```
Comment (stephanosio) [inline boards/posix/native_posix/Kconfig.defconfig:19]: Use TAB

**Review by stephanosio — DISMISSED** [review]

**Comment by aescolar** [comment]
@yperess Would you be so kind as to elaborate on the motivation?

**Comment by yperess** [comment]
Sure, I'm writing tests for a math lib in chromium and the job supports both fixed and floating point. I need to enable the FPU in native_posix boards in order to run the tests.

File: boards/posix/native_posix/Kconfig.defconfig (hunk 1 above)
Comment (yperess) [reply boards/posix/native_posix/Kconfig.defconfig:19]: Done

**Review by stephanosio — COMMENTED** [review]
FWIW, `CPU_HAS_FPU` is a "SoC config" and is almost always selected by the SoC symbol; so, in this case, `SOC_POSIX` should `select CPU_HAS_FPU`.

https://github.com/zephyrproject-rtos/zephyr/blob/6f5bd8d963e85767fb4b0395b4c76d8686b8e005/soc/posix/inf_clock/Kconfig.soc#L3

Of course, the POSIX arch is a special case and one could argue these conventions do not apply, but I see little reason to not follow the existing convention in this particular case.

If being able to de-select `CPU_HAS_FPU` for the POSIX boards was a requirement for whatever reason, then `SOC_POSIX` can `imply CPU_HAS_FPU`.

**Comment by yperess** [comment]
> select CPU_HAS_FPU

Thanks, that works for my tests. I left it as `select` vs `imply` because I do want to run the tests with FPU both on and off on the CI.

**Review by stephanosio — DISMISSED** [review]

**Review by aescolar — CHANGES_REQUESTED** [review]
> Of course, the POSIX arch is a special case and one could argue these conventions do not apply, but I see little reason to not follow the existing convention in this particular case.

~In this regard, the POSIX arch/soc/board layering is very different as the "soc" does not provide anything a real soc provides and all peripherals are defined at the board level.
One simulated "board" may want to reflect a real SOC having an FPU or not if the code will change depending on this.
But similarly to how you wanted to be able to enable the FPU for your tests in native_posix some others may want to disable it. So please move it to the board level, but allow users to override it.~

**Comment by stephanosio** [comment]
> But similarly to how you wanted to be able to enable the FPU for your tests in native_posix some others may want to disable it. So please move it to the board level, but allow users to override it.

Enabling FPU should be done through `CONFIG_FPU` and not through `CONFIG_CPU_HAS_FPU`; the latter merely specifies whether the SoC, or in this case the host system, is capable of executing floating-point instructions which is always the case for x86(-64) and AArch64 Linux hosts.

Semantically `CONFIG_CPU_HAS_FPU` should not be overridable because it is more of a "property" than a "configuration," but if one wants to do that for whatever reason, the SoC symbol can `imply` instead of `select`.

**Review by aescolar — DISMISSED** [review]

**Comment by aescolar** [comment]
> Enabling FPU should be done through `CONFIG_FPU` and not through `CONFIG_CPU_HAS_FPU`

You are very right

**Comment by yperess** [comment]
Rebased to get the CI to pass

**Review by stephanosio — DISMISSED** [review]

**Comment by yperess** [comment]
@nashif it seems that there's an alignment error unrelated to this change that's still failing the CI. The reviews were on the most recent commit, I simply rebased hoping that the issue was fixed in `main` already.

**Comment by yperess** [comment]
@stephanosio this CI failure is interesting (I thought it was unrelated at first). It looks like the filter is for `integration_platforms` set to `native_posix` but then tests like `libraries.cbprintf_package_fp` set a filter for having the FPU which would exclude `native_posix`. Does that mean that `libraries.cbprintf_package_fp` was never run?

**Comment by yperess** [comment]
> https://github.com/zephyrproject-rtos/zephyr/blob/6a0df63ea0a5cf2029d8d74a0ec46ed35d668c49/tests/lib/cbprintf_package/testcase.yaml#L19-L20
> 
> > Does that mean that `libraries.cbprintf_package_fp` was never run?
> 
> Yes, that would have been the case.
> 
> @nordic-krch any suggestions?

I think I found a fix (worked for me locally).

**Comment by yperess** [comment]
OK, the latest set of changes seems to pass all the tests that I'm able to also pass on ToT.

**Comment by yperess** [comment]
@stephanosio @aescolar I believe this is ready now. The documentation build is broken, but it looks unrelated to this change. Turns out this uncovered a few bugs so I'm overall really happy with this change.

**Comment by nordic-krch** [comment]
Changes look ok. Not sure what is wrong with Doc build.

**Comment by stephanosio** [comment]
@gmarull any idea about the doc build failure?

**Comment by gmarull** [comment]
> @gmarull any idea about the doc build failure?

There's something Sphinx C domain doesn't like:

```
Traceback (most recent call last):                                                                                                                        
  File "/home/gmarull/ws/nordic/zephyrproject/.venv/lib/python3.9/site-packages/sphinx/domains/c.py", line 3263, in handle_signature                      
    parser.assert_end()                                                                                                                                   
  File "/home/gmarull/ws/nordic/zephyrproject/.venv/lib/python3.9/site-packages/sphinx/util/cfamily.py", line 344, in assert_end                          
    self.fail('Expected end of definition.')                                                                                                              
  File "/home/gmarull/ws/nordic/zephyrproject/.venv/lib/python3.9/site-packages/sphinx/util/cfamily.py", line 275, in fail                                
    raise self._make_multi_error(errors, '')                                                                                                              
sphinx.util.cfamily.DefinitionError: Invalid C declaration: Expected end of definition. [error at 17]                                                     
  cbprintf package alignment must be a power of                                                                                                           
  -----------------^        
```

I'll try to investigate what's going on here... (trackeback can be obtained by adding `-T` to `SPHINXOPTS`).

**Comment by yperess** [comment]
> > @gmarull any idea about the doc build failure?
> 
> There's something Sphinx C domain doesn't like:
> 
> ```
> Traceback (most recent call last):                                                                                                                        
>   File "/home/gmarull/ws/nordic/zephyrproject/.venv/lib/python3.9/site-packages/sphinx/domains/c.py", line 3263, in handle_signature                      
>     parser.assert_end()                                                                                                                                   
>   File "/home/gmarull/ws/nordic/zephyrproject/.venv/lib/python3.9/site-packages/sphinx/util/cfamily.py", line 344, in assert_end                          
>     self.fail('Expected end of definition.')                                                                                                              
>   File "/home/gmarull/ws/nordic/zephyrproject/.venv/lib/python3.9/site-packages/sphinx/util/cfamily.py", line 275, in fail                                
>     raise self._make_multi_error(errors, '')                                                                                                              
> sphinx.util.cfamily.DefinitionError: Invalid C declaration: Expected end of definition. [error at 17]                                                     
>   cbprintf package alignment must be a power of                                                                                                           
>   -----------------^        
> ```
> 
> I'll try to investigate what's going on here... (trackeback can be obtained by adding `-T` to `SPHINXOPTS`).

It seems that it just doesn't like my assert error message, should I just replace `"cbprintf package alignment must be a power of 2"` with something else? I think even `NULL` will be fine as the assert itself is pretty descriptive.

**Comment by yperess** [comment]
OK, looks like `NULL` isn't a valid assert :), using plain `""` now should pass the CI

**Comment by stephanosio** [comment]
And this pops up all of sudden ...

```
Warning, treated as error:
/home/runner/work/zephyr/zephyr/doc/_build/src/releases/release-notes-3.0.rst:6[55](https://github.com/zephyrproject-rtos/zephyr/runs/5636142328?check_suite_focus=true#step:7:55):Unknown interpreted text role "kconfig".
```

**Comment by yperess** [comment]
> And this pops up all of sudden ...
> 
> ```
> Warning, treated as error:
> /home/runner/work/zephyr/zephyr/doc/_build/src/releases/release-notes-3.0.rst:6[55](https://github.com/zephyrproject-rtos/zephyr/runs/5636142328?check_suite_focus=true#step:7:55):Unknown interpreted text role "kconfig".
> ```
> 
> p.s. https://github.com/zephyrproject-rtos/zephyr/actions/workflows/doc-build.yml
> 
> At least, it is not an isolated issue in this PR; though, it started appearing globally after the last doc build for this PR which I am inclined to believe is just by pure chance (~maybe one of the doc dependencies got updated causing this?~) ...
> 
> UPDATE: It was indeed by pure chance ... main is broken: [fff22d3](https://github.com/zephyrproject-rtos/zephyr/commit/fff22d3d7f03f604f104bfaef30b373d30ef9ce8)

Thanks, so is this PR good to merge?

File: include/sys/cbprintf.h [hunk 2]
```diff
@@ -49,22 +49,25 @@ extern "C" {
  * @{
  */
 
-/** @brief Required alignment of the buffer used for packaging. */
+#define POWER_OF_TWO(x) ((x) && !((x) & ((x) - 1)))
```
Comment (stephanosio) [inline include/sys/cbprintf.h:52]: Generally, local macro definitions like these should be avoided to prevent conflicts; if it is necessary, the relevant toolchain or util headers should be updated to support them.

File: include/sys/cbprintf.h [hunk 3]
```diff
@@ -49,22 +49,25 @@ extern "C" {
...
+#define __CBPRINTF_PACKAGE_ALIGNMENT 16
 #else
-#define CBPRINTF_PACKAGE_ALIGNMENT \
+#define __CBPRINTF_PACKAGE_ALIGNMENT \
 	(IS_ENABLED(CONFIG_CBPRINTF_PACKAGE_LONGDOUBLE) ? \
 		sizeof(long double) : MAX(sizeof(double), sizeof(long long)))
 #endif
 
+/** @brief Required alignment of the buffer used for packaging. */
+#ifdef Z_POW2_CEIL
```
Comment (stephanosio) [inline include/sys/cbprintf.h:63]: Why is this required? If the toolchain abstraction layer defines `Z_POW2_CEIL`, it should be available for all supported toolchains.

File: include/sys/cbprintf.h (hunk 2 above)
Comment (yperess) [reply include/sys/cbprintf.h:52]: Done

File: include/toolchain/gcc.h [hunk 4]
```diff
@@ -591,10 +591,18 @@ do {                                                                    \
...
+		1UL << (31U - __builtin_clz(x)))
 #endif
 
+/**
+ * @brief Check whether or not a value is a power of 2
+ *
+ * @param x The value to check
+ * @return true if x is a power of 2, false otherwise
+ */
+#define Z_IS_POW2(x) ((x) && !((x) & ((x) - 1)))
```
Comment (stephanosio) [inline include/toolchain/gcc.h:605]: It looks like the Xtensa arch code defines its own `Z_IS_POW2` -- we should remove that.

https://github.com/zephyrproject-rtos/zephyr/blob/890f6199367953e52f2ffcf10420c98d9ce3a320/include/arch/xtensa/cache.h#L18

Also, to make it more compliant to the Zephyr/MIRSA coding guidelines:
```suggestion
#define Z_IS_POW2(x) (((x) != 0) && (((x) & ((x) - 1U)) == 0))
```

p.s. the global addition of `Z_IS_POW2` should be in a separate commit.

**Review by stephanosio — COMMENTED** [review]
This PR is turning out to be much more work than anticipated. Thanks for your patience.

File: include/toolchain/gcc.h (hunk 4 above)
Comment (yperess) [reply include/toolchain/gcc.h]: Done

File: include/sys/cbprintf.h (hunk 3 above)
Comment (yperess) [reply include/sys/cbprintf.h]: I removed this and rebased, but now I can't seem to get the tests to pass anymore, will keep debugging...

**Comment by yperess** [comment]
> This PR is turning out to be much more work than anticipated. Thanks for your patience.

No worries at all. I'm surprised how much work just turning on the FPU was, but I'm glad it seems to be working now.

**Review by stephanosio — APPROVED** [review]

**Review by aescolar — APPROVED** [review]
+1
